Qa python homework 2

bogol243 opened the pull request:
# Домашнее задание №

Прошу @snicks92 или @z1ll1 или @SK1995 проверить его.

Что было сделано:
Настроено окружение (установлен докер, селеноид)
Зарегистрирован марк UI
Написаны тесты на авторизацию
Написан тест на создание рекламной кампании
Написан тест на создание сегмента
Написан тест на удаление сегмента после создания (1 балл)
Написана фикстура, возвращающая объект главной страницы после авторизации

bogol243 committed:
Update .gitignore

bogol243 committed:
Update .gitignore

bogol243 committed:
Update requirements.txt

bogol243 committed:
Create browsers.json

bogol243 committed:
add homework files

bogol243 committed:
removed chrome_ver from settings and imports

SK1995 commented on tests_ui/tests/tests.py:
```diff
@@ -0,0 +1,96 @@
+    try:
+        dashboard_page.find_campaign()
+    except TimeoutException:
+        assert False
```
Выглядит как-то не очень.
Зачем делать `assert False`, если это явно будет бросать исключение?
Причем в тесте лог об ошибке так и будет выведен - `Assertion error: assert False`, и по нему будет очень весело искать причину ошибки :)

Думаю, тут лучше в самом методе `find_campaign` возвращать либо `None`, если кампания не найдена, либо `WebElement` с ней, если найдена. И тогда в тесте вместо этого можно будет делать просто вот так:
```python
campaign = dashboard_page.find_campaign()
assert campaign is not None
```

SK1995 commented on tests_ui/tests/tests.py:
```diff
@@ -0,0 +1,96 @@
+    try:
+        page.find_segment_to_delete()
+    except TimeoutException:
+        assert True
```
Тоже немного странная конструкция.
Она не помогает читабельности кода, а скорее наоборот.
Возникает вопрос, зачем делать `assert True`, если это всегда верно.

SK1995 commented on tests_ui/tests/tests.py:
```diff
@@ -0,0 +1,96 @@
+        page.find_segment_to_delete()
+    except TimeoutException:
+        assert True
+        return
```
В тестах не принято делать `return`.
Они должны согласно своему коду завершаться.

SK1995 commented on tests_ui/ui/settings.py:
```diff
@@ -0,0 +1,7 @@
+
+credentials_wrong = {"email": "aaaaaaa@aaa.com","password":"xxxxxxxxxxxx"}
+
+campaign_name = "test ad campaign"
```
Лучше в тестах к этому имени добавлять хоть немного уникальности.
Например, через `str(uuid.uuid4())`
Сейчас все тесты, получается, используют одинаковые имена для кампаний и сегментов, это может привести к тому, что тесты начнут друг друга аффектить при параллельном запуске.

SK1995 reviewed (changes requested):
Спасибо, хорошая работа!

Оставил несколько комментариев по коду, которые было бы неплохо поправить.

Отдельно хочется добавить про сложности с запуском тестов:
1) tests.py - не коллектится автоматически, тк в имени фала нет префикса `test_`. Как ты проверял работу тестов-то?
2) При запуске из папки с новыми тестами (tests_ui) ошибка `No module named 'tests_ui'`
3) Тесты не запускаются вне селеноида из-за проблем с версией хрома ( в селеноиде и в webdriver-manager они отличаются)

Эти замечания можно уже не исправлять, но в следующий раз прошу на них обратить внимание. Мне пришлось помучиться, чтобы запустить код для проверки :)

bogol243 commented:
Привет! Спасибо за проверку. Комментарии посмотрел, понял, ошибки учту, исправлю. Тесты запускал передавая пайтесту файл в параметрах, исправлю это тоже.
Насчёт второго пункта, да у меня vscode говорит что не может заимпортить, но при этом видит и импортит:) У меня в настройках прописан pythonPath, поэтому всё и работает. Как можно решить проблему с импортами модулей чтобы работало не только у меня, а везде?
Насчёт версии браузера, да я по умолчанию поставил 80.0 в соответствии с заданием. На своём компьютере, для отладки, запускал с --browser_ver latest.

bogol243 committed:
create fixture campaign_name

test_create_campaign test now uses unique value from campaign_name fixture instead of fixed

bogol243 committed:
Improved readability of test_create_campaign()

DashboardPage find_campaign() now hanldles exceptions by itself. And returnes WebElement or None if TimeoutException has occured

bogol243 committed:
Improved readability of test_delete_segment()

SegmentsPage find_element_to_delete() now handles exceptions by itself. Returns WebElement if it's found and None if TimeoutException has occured.

bogol243 committed:
improved readability of test_create_segment()

bogol243 committed:
Random name of created segment

Name of created segment now unique and random, therefore locators which used this name are now functions returning the actual locator

bogol243 committed:
Useless condition deleted

SK1995 commented:
Спасибо, стало еще немножко лучше!
Поставил максимум (12 баллов).

По поводу того, как решить проблему с импортами - обычно подразумевается, что тесты запускаются из корневого каталога с новым кодом домашнего задания (в данном случае это `tests_ui`). Будет достаточно сделать так, чтобы оттуда командой `pytest -s -l -v ./` происходил запуск тестов.
Но вообще я в домашнем задании ничего об этом не написал, так что тут точно не буду придираться, тем более что работа правда хорошая :)
